Check next update and cache CRLs in the X509 authenticator

rmartinc opened the pull request:
Closes #35983
Closes #26473

Draft for the moment. The PR is separated into two commits because probably only the first one is back-ported to the previous releases. The idea is that now CRLs are cached until the next update time or the specified cache time in the provider (the minimum of both). Besides there is a configuration parameter to the X509 authenticator that allows using non-updated CRLs to maintain previous behavior. Decisions to consider:

1. The configuration option in X509 is a breaking change in compatibility. In order to work as before the option `x509-cert-auth-crl-continue-if-non-updated` should be set to `true` instead of the default `false`. I know we should be backwards compatible, but this is a security option and I think we should be secure by default and add a migration note. So I'm requesting an exception for this PR (we can do the other way around if we want to maintain backwards compatibility but not be secure by default).
2. A new `crl` cache is added following the idea in the old PR linked in the previous point. This cache is always local (copied from the `keys` cache but without expiration) and the X509CRL object is cached directly into it. This should be checked by the sre team for sure :smile:. 
3. The main difference to the previous PR #25863 is that the full X509CRL is cached instead of using BC to parse and load it into memory. Depending the JCE provider used in [this line](https://github.com/keycloak/keycloak/blob/26.0.7/services/src/main/java/org/keycloak/authentication/authenticators/x509/CertificateValidator.java#L279) (normally SUN in normal configuration but it can be configured different using the security JDK file, BCFIPS in fips configuration for sure), the behavior is different. In the default SUN implementation the CRL entries are cached into a map keyed by the serial number. The size and performance is similar to the one implemented by the streaming parser provided in the old PR. In the BC (and BCFIPS) implementation the CRL is not cached and a call to the `isRevoked` method iterates the CRL entries. So SUN implementation is big in memory because of the map but fast. BC/BCFIPS impl is small in size but slow in performance for very long CRLs. We can improve this. My vote would be always using the provider obtained by the `CryptoIntegration` (BC or BCFIPS) and add the map to the cache entry implementing our own X509CRL wrapping the one provided by the BC provider (If cache is in place I prefer using more memory but process the list faster). But I want to manage this as a follow-up issue to not add to many changes in this PR.
4. Note that now with the crl cache only one X509CRL object is used for each url reducing the memory used. This object is accessed concurrently but read-only by each thread.

rmartinc committed:
Add crl cache to certificate validation
Closes #26473

Signed-off-by: rmartinc <rmartinc@redhat.com>

rmartinc committed:
Check next update time for CRL in certificate validation
Closes #35983

Signed-off-by: rmartinc <rmartinc@redhat.com>

mposolda reviewed (commented):
@rmartinc Great stuff, Thanks!

- Do we need also the ability to "clear crls" ? I see that most of other entities like realms, users and keys have it (For example see `ClearKeysCacheResource` and related `CachePublicKeyProvider`). At the same time, I am not sure if this is strictly needed as it seems that those admin endpoints for clearing caches cannot be triggered from admin console (unless I am missing something). Maybe we can have a follow-up issue for clearing support if it makes sense (unless it makes it worse for backports if we need to backport this). Feedback from the SRE team would be welcome (CC @ahus1 )


- Backwards compatibility and the option `continueIfNonUpdated` : I appreciate the security, however the breaking change can be problematic as the setups, which were working before, may now stop working. And since this is not CVE, it can be good to avoid breaking changes IMO. Is it possible to make it in a way that:
  - For new deployments, it would be `false` by default. So when someone creates the X.509 authentication provider in the admin console, it would be pre-selected as `false`
  - For existing deployments, it would be `true` by default. So when migrating DB or importing realm from the previous version, it would be considered `true` for backwards compatibility (specifically when the option does not exists).

Is it ok to you? If it makes things easier, I am fine with renaming the option and make it the opposite way (so `true` for new deployments and `false` for existing deployments when `false` would mean that expiration is ignored). Whatever you prefer.


- Thanks for the note around your point (3) ! +1 for handle (3) as the follow-up issue.

- I suppose we need something in the release notes (for the new option and for the new cache as well)? But it can make it harder for backport as I guess the release notes would be different if we backport just the authenticator option, but not new infinispan cache to the older releases like 26.0 or 24.0. But hopefully it is solvable by manual updating release notes in case backports are needed?

mposolda commented:
@ahus1 Can you please review or delegate to someone from SRE team to review? The PR introduces new cache (very similar to `keys` cache). We would be happy for the feedback.

ahus1 commented:
@pruivo - can you please review this new cache from an Infinispan perspective as asked in https://github.com/keycloak/keycloak/pull/36294#issuecomment-2598380541? Thanks!

ahus1 reviewed (dismissed):
Please update https://github.com/keycloak/keycloak/blob/caf546bdb2a117240739f6592340a557ded51128/docs/guides/server/caching.adoc?plain=1#L35-L51 to list this new cache. 

For other local caches, we allow configuring the size on the CLI. Not sure how relevant it is here, still let's be consistent. 

https://github.com/keycloak/keycloak/blob/4a924f6c944fc75ca06a371f9d5305ef6fbc3301/quarkus/config-api/src/main/java/org/keycloak/config/CachingOptions.java#L29

rmartinc commented:
@ahus1! Done! Let's see if adding the new option does not break any test... :smile:

ahus1 commented:
@rmartinc - usually these changes to the CLI break the help command tests, as the output changes. Let's see. 

BTW, after my last comment I found that there is some "clear cache" functionality in the UI, actually in "Authentication" (not sure why). See below for a screenshot. This doesn't list the new cache. I don't know where in the UI you would clear the new cache for CRLs.

![image](https://github.com/user-attachments/assets/416607d2-7347-4dd8-be13-76b47947263c)

rmartinc commented:
OK! I fixed the texts for the new cache entry to fix the CI failures. I have also added the changes in the admin console (I created the endpoint but I didn't realize that there was a page to clear the caches in the new console, I thought that this functionality was removed, but it seems that it was re-added at some point :stuck_out_tongue:). Let's see if everything works now.

pruivo commented on model/infinispan/src/main/java/org/keycloak/crl/infinispan/InfinispanCrlStorageProvider.java:
```diff
@@ -0,0 +1,135 @@
+            int cacheTime, int minTimeBetweenRequests) {
+        this.crlCache = crlCache;
+        this.tasksInProgress = tasksInProgress;
+        this.cacheTime = cacheTime > 0? TimeUnit.SECONDS.toMillis(cacheTime) : -1;
+        this.minTimeBetweenRequests = minTimeBetweenRequests > 0? TimeUnit.SECONDS.toMillis(minTimeBetweenRequests) : 10000L;
```
I would do this computation in `InfinispanCrlStorageProviderFactory.init(Config.Scope config)` to avoid computing the same things over and over again.

pruivo commented on model/infinispan/src/main/java/org/keycloak/crl/infinispan/InfinispanCrlStorageProvider.java:
```diff
@@ -0,0 +1,135 @@
+
+    private static final Logger log = Logger.getLogger(InfinispanCrlStorageProvider.class);
+
+    private final Cache<String, X509CRLEntry> crlCache;
+    private final Map<String, FutureTask<X509CRL>> tasksInProgress;
+    private final long cacheTime;
+    private final long minTimeBetweenRequests;
```
If you want to reduce memory allocation, create an interface to expose this shared instances. Similar to what I did here

https://github.com/keycloak/keycloak/blob/9006218559496d5fb6fe98cd9760cad7f57b62e7/model/infinispan/src/main/java/org/keycloak/cluster/infinispan/remote/RemoteInfinispanClusterProvider.java#L47

https://github.com/keycloak/keycloak/blob/9006218559496d5fb6fe98cd9760cad7f57b62e7/model/infinispan/src/main/java/org/keycloak/cluster/infinispan/remote/RemoteInfinispanClusterProvider.java#L161-L166

pruivo commented on model/infinispan/src/main/java/org/keycloak/crl/infinispan/InfinispanCrlStorageProvider.java:
```diff
@@ -0,0 +1,135 @@
+        } catch (ExecutionException ee) {
+            throw new RuntimeException("Error when loading crl " + key + " : " + ee.getMessage(), ee);
+        } catch (InterruptedException ie) {
+            throw new RuntimeException("Error. Interrupted when loading crl " + key, ie);
```
restore thread interrupt signal
```suggestion
           Thread.currentThread().interrupt();
            throw new RuntimeException("Error. Interrupted when loading crl " + key, ie);
```

pruivo commented on model/infinispan/src/main/java/org/keycloak/crl/infinispan/InfinispanCrlStorageProvider.java:
```diff
@@ -0,0 +1,135 @@
+            log.warnf("Loading crl with key '%s' returned null.", key);
+            return null;
+        }
+        long lifespan = cacheTime;
+        if (crl.getNextUpdate() != null) {
+            final long nextUpdateTime = crl.getNextUpdate().getTime() - currentTime;
+            if (nextUpdateTime <= 0) {
+                // if the CRL is expired just cache the minimum time
+                lifespan = minTimeBetweenRequests;
+            } else if (lifespan <= 0 || nextUpdateTime < lifespan) {
+                // get the minimum between cacheTime and nextUpdate
+                lifespan = nextUpdateTime;
+            }
+        }
```
extract to a method to improve code readability? My suggestion
```
long lifespan = getLifespan(crl, currentTime);
```
```
    private long getLifespan(X509CRL crl, long currentTime) {
        if (crl.getNextUpdate() == null) {
            return cacheTime;
        }
        final long nextUpdateTime = crl.getNextUpdate().getTime() - currentTime;
        if (nextUpdateTime <= 0) {
            // if the CRL is expired just cache the minimum time
            return minTimeBetweenRequests;
        }
        // get the minimum between cacheTime and nextUpdate
        return Math.min(cacheTime, nextUpdateTime);
    }

```

pruivo replied:
nice catch :+1:

pruivo commented on model/infinispan/src/main/java/org/keycloak/crl/infinispan/InfinispanCrlStorageProvider.java:
```diff
@@ -0,0 +1,135 @@
+    }
+
+    @Override
+    public boolean refreshCache(String key, Callable<X509CRL> loader) throws GeneralSecurityException {
```
`refreshCache` is not used. Assuming it will be used in the future?

rmartinc replied:
I just did it because it is done in keys. In keys it is used to manually refresh the cache in the admin UI. I have maintained it but we can remove it for now if you prefer to not have it.

pruivo replied:
No preference from me.

pruivo commented on model/infinispan/src/main/java/org/keycloak/crl/infinispan/InfinispanCrlStorageProviderFactory.java:
```diff
@@ -0,0 +1,115 @@
+                    .helpText(
+                            """
+                            Interval in seconds that the CRL is cached. The next update time of the CRL is always a minimum if present.
+                            Zero or a negative value means CRL is cached until the next update time specified in the CRL (or inifinite if the
```
typo
```suggestion
                            Zero or a negative value means CRL is cached until the next update time specified in the CRL (or infinite if the
```

pruivo commented on model/infinispan/src/main/java/org/keycloak/crl/infinispan/InfinispanCrlStorageProviderFactory.java:
```diff
@@ -0,0 +1,115 @@
+                            Minimum interval in seconds between two requests to retrieve the CRL. The CRL is not updated
+                            from the URL again until this minimum time has passed since the previous refresh. In theory
+                            this option is never used if the CRL is refreshed correctly in the next update time.
+                            Default 10 seconds.
```
is a zero or a negative value valid? Probably worth adding a sentence about the behavior with a zero or negative value.

pruivo reviewed (changes requested):
The copyright year is incorrect in some new classes. I focused mostly on the Infinispan providers and added some suggestions. Thanks!

rmartinc commented:
@pruivo I have added all your suggestions except the `refreshCache`, that I maintained it. I have no strong opinion on it, so I can remove it if it's preferred by everybody. It was just added because it was present in the keys cache (to be used by the admin UI), but it's not used for the moment, so we can remove it without any problem.

rmartinc commented on model/infinispan/src/main/java/org/keycloak/crl/infinispan/X509CRLEntry.java:
```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2024 Red Hat, Inc. and/or its affiliates
```
I have changed all new years to 2025. I started this on 2024... :smile:

pruivo commented on services/src/main/java/org/keycloak/authentication/authenticators/x509/CertificateValidator.java:
```diff
@@ -261,54 +263,71 @@ public static class CRLFileLoader extends CRLLoaderImpl {
 
         public Collection<X509CRL> getX509CRLs() throws GeneralSecurityException {
-            CertificateFactory cf = CertificateFactory.getInstance("X.509");
-            Collection<X509CRL> crlColl = null;
+            CrlStorageProvider crlCache = session.getProvider(CrlStorageProvider.class);
+            X509CRL crl = null;
+            if (cRLPath != null) {
+                crl = crlCache.get(cRLPath, this::loadCRL);
+            }
+
+            if (crl == null) {
+                throw new GeneralSecurityException(String.format("Unable to load CRL from \"%s\"", cRLPath));
+            }
+
+            if (crl.getNextUpdate() != null && crl.getNextUpdate().compareTo(new Date(Time.currentTimeMillis())) < 0) {
+                final String message = String.format("CRL from '%s' is not refreshed. Next update is %s.", cRLPath, crl.getNextUpdate());
+                logger.warn(message);
+                if (abortIfNonUpdated) {
+                    throw new GeneralSecurityException(message);
+                }
+            }
 
+            return Collections.singletonList(crl);
+        }
```
Optionally, simplify if chains. IMO, checking if `cRLPath` is null should be the first thing to avoid lookup for the `CrlStorageProvider`
```java
        public Collection<X509CRL> getX509CRLs() throws GeneralSecurityException {
            if (cRLPath == null) {
                throw new GeneralSecurityException("Unable to load CRL");
            }

            X509CRL crl = session.getProvider(CrlStorageProvider.class).get(cRLPath, this::loadCRL);
            if (crl == null) {
                throw new GeneralSecurityException(String.format("Unable to load CRL from \"%s\"", cRLPath));
            }

            if (crl.getNextUpdate() == null) {
                return Collections.singletonList(crl);
            }

            if (crl.getNextUpdate().getTime() < Time.currentTime()) {
                final String message = String.format("CRL from '%s' is not refreshed. Next update is %s.", cRLPath, crl.getNextUpdate());
                logger.warn(message);
                if (abortIfNonUpdated) {
                    throw new GeneralSecurityException(message);
                }
            }

            return Collections.singletonList(crl);
        }
```

pruivo commented on services/src/main/java/org/keycloak/authentication/authenticators/x509/CertificateValidator.java:
```diff
@@ -261,54 +263,71 @@ public static class CRLFileLoader extends CRLLoaderImpl {
+        }
+
+        private X509CRL loadCRL() throws GeneralSecurityException {
+            CertificateFactory cf = CertificateFactory.getInstance("X.509");
```
check if `cRLPath` is null first to avoid lookups.

rmartinc replied:
I just removed the if here as it was done before in the previous method.

pruivo commented on services/src/main/java/org/keycloak/authentication/authenticators/x509/CertificateValidator.java:
```diff
@@ -261,54 +263,71 @@ public static class CRLFileLoader extends CRLLoaderImpl {
         }
 
-        private Collection<X509CRL> loadFromURI(CertificateFactory cf, URI remoteURI) throws GeneralSecurityException {
+        private X509CRL loadFromURI(CertificateFactory cf, URI remoteURI) throws GeneralSecurityException {
```
Looks like `loadFromURI`, `loadCRLFromLDAP`, `loadFromURI`, and `loadFromStream` can be static methods as they do not use any fields.

rmartinc replied:
Not all of them, `loadFromUri` uses the session and `loadCRLFromLDAP` uses ldapContext. So I have not changed any of them.

pruivo reviewed (approved):
LGTM @rmartinc 
Took a glance at the remaining code and have some suggestions in place. All of them are optional and no objection to me merging this PR as is. 
Thanks!

pruivo commented on services/src/main/java/org/keycloak/authentication/authenticators/x509/CertificateValidator.java:
```diff
@@ -374,15 +393,15 @@ private Collection<X509CRL> loadCRLFromFile(CertificateFactory cf, String relati
-            return Collections.emptyList();
+            return null;
         }
         private X509CRL loadFromStream(CertificateFactory cf, InputStream is) throws IOException, CRLException {
```
Ops, one comment was no save. It seems you don't need to wrap the `InputStream` into a `DataInputStream`. 
```java
        private static X509CRL loadFromStream(CertificateFactory cf, InputStream is) throws IOException, CRLException {
            return (X509CRL) cf.generateCRL(is);
        }

```

rmartinc replied:
Done! I have also checked all of them closes the InputStream with a try.

rmartinc commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/x509/CRLRule.java:
```diff
@@ -50,17 +52,25 @@ public class CRLRule extends ExternalResource {
-        pathHandler.addExactPath(AbstractX509AuthenticationTest.INTERMEDIATE_CA_INVALID_SIGNATURE_CRL_PATH, new CRLHandler(AbstractX509AuthenticationTest.INTERMEDIATE_CA_INVALID_SIGNATURE_CRL_PATH));
-        pathHandler.addExactPath(AbstractX509AuthenticationTest.INTERMEDIATE_CA_3_CRL_PATH, new CRLHandler(AbstractX509AuthenticationTest.INTERMEDIATE_CA_3_CRL_PATH));
-        pathHandler.addExactPath(AbstractX509AuthenticationTest.INVALID_CRL_PATH, new CRLHandler(AbstractX509AuthenticationTest.INVALID_CRL_PATH));
+        handlers = new HashMap<>();
```
I added the method later to do some tests and I didn't detect it can be used. Nice!

rmartinc commented:
OK. I think that now it's OK. So I will prepare the release notes and documentation. But first I want to confirm what is going to be backported and what not. I will check in today's meeting and I will send the documentation changes to move this out of draft.

rmartinc commented:
I have rebased because there was a conflict with the release notes. If the tests are OK I will move this out of draft.

mposolda commented on docs/documentation/upgrading/topics/changes/changes-26_0_10.adoc:
```diff
@@ -0,0 +1,5 @@
+= New option in X.509 authenticator to abort authentication if CRL is outdated
```
Few points to the documentation:

- It seems this is backwards compatible change for most of the cases, so it likely belongs more to the release notes rather than to the upgrading guide? AFAIK the strategy changed and from 26.1.0, the upgrading guide contains only 2 things, which may affect behaviour for backwards compatibility (offline sessions, service accounts). This one won't affect backwards compatibility, so likely also the release notes?

- Do we want to backport this commit for this option to 26.1 as well? I suppose that yes? In that case, it should be documented in the release notes in the 26.1.1 rather than 26.0.10. My thinking is this: When someone upgrades from `26.1.0` to `26.1.1`, they will read just the release notes for that particular version. They won't read the release notes (or upgrading guide) of the 26.0.10 release.

Considering this, I assume that:
- In branches `main` and in `release/26.1`, we want to document this in the release notes of 26.1.0
- In `release/26.0`, this should go in the release notes of 26.0.10 . So people upgrading to RHBK 26.0.10 will notice this.

andymunro replied:
I agree with this logic, @mposolda

mposolda reviewed (commented):
@rmartinc Thanks for the updates and for the documentation!

Docs itself looks great to me. I have just some points related where to place the docs for the new option of X509 authenticators (in shortcut, I believe that in `main` it should be 26.1.1 release notes rather than 26.0.10 upgrading guide. More details in the inline comment).

@andymunro Do you please have a chance to review the documentation?

andymunro commented on docs/documentation/server_admin/topics/authentication/x509.adoc:
```diff
@@ -134,6 +134,9 @@ Use CDP to check the certificate revocation status. Most PKI authorities include
 The path to a file containing a CRL list. The value must be a path to a valid file if the *CRL Checking Enabled* option is enabled.
 
+*CRL abort if non updated*::
+A CRL conforming link:https://datatracker.ietf.org/doc/html/rfc5280#section-5.1.2.5[RFC5280] contains a next update field that indicates the date by which the next CRL will be issued. When that time is passed, the CRL is considered outdated and it should be refreshed. If this option is `true`, the authentication will fail if the CRL is outdated (recommended). If the option is set to `false`, the  ...
```
```suggestion
A CRL conforming to link:https://datatracker.ietf.org/doc/html/rfc5280#section-5.1.2.5[RFC5280] contains a next update field that indicates the date by which the next CRL will be issued. When that time is passed, the CRL is considered outdated and it should be refreshed. If this option is `true`, the authentication will fail if the CRL is outdated (recommended). If the option is set to `false`, the outdated CRL is still used to validate the user certificates.
```

andymunro commented on docs/documentation/upgrading/topics/changes/changes-26_0_10.adoc:
```diff
@@ -0,0 +1,5 @@
+= New option in X.509 authenticator to abort authentication if CRL is outdated
+
+The X.509 authenticator has a new option `x509-cert-auth-crl-abort-if-non-updated` (*CRL abort if non updated* in the admin console) to abort the login if a CRL is configured to validate the certificate and the CRL is not updated in the time specified in the next update field. The new option defaults to `true` in the admin console. For more details about the CRL next update field, check link:http ...
```
```suggestion
The X.509 authenticator has a new option `x509-cert-auth-crl-abort-if-non-updated` (*CRL abort if non updated* in the Admin Console) to abort the login if a CRL is configured to validate the certificate and the CRL is not updated in the time specified in the next update field. The new option defaults to `true` in the Admin Console. For more details about the CRL next update field, see link:https://datatracker.ietf.org/doc/html/rfc5280#section-5.1.2.5[RFC5280, Section-5.1.2.5].
```

andymunro commented on docs/documentation/upgrading/topics/changes/changes-26_0_10.adoc:
```diff
@@ -0,0 +1,5 @@
+
+The X.509 authenticator has a new option `x509-cert-auth-crl-abort-if-non-updated` (*CRL abort if non updated* in the admin console) to abort the login if a CRL is configured to validate the certificate and the CRL is not updated in the time specified in the next update field. The new option defaults to `true` in the admin console. For more details about the CRL next update field, check link:http ...
+
+The value `false` is maintained for compatibility with the previous behavior. Note that existing configurations will not have the new option and will act as if this option was set to `false`, but the admin console will add the default value `true` on edit.
```
```suggestion
The value `false` is maintained for compatibility with the previous behavior. Note that existing configurations will not have the new option and will act as if this option was set to `false`, but the Admin Console will add the default value `true` on edit.
```

andymunro commented on quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelp.approved.txt:
```diff
@@ -32,6 +32,8 @@ Cache:
                        clientSessions cache. Available only when embedded Infinispan clusters
                        configured.
+--cache-embedded-crl-max-count <max-count>
+                     The maximum number of entries that can be stored in-memory by the crl cache.
```
```suggestion
                     The maximum number of entries that can be stored in-memory by the crl cache
```

andymunro replied:
We now omit periods if it is not a full sentence.

rmartinc replied:
@andymunro all this files are generated automatically with the cache names. So in order to change this one I have to change the messages for all the caches and also change all the files. I prefer to not do this here. Besides I see all the meesages in the help display finish with a point:

```
./kc.sh start-dev --help
...

--cache-embedded-client-sessions-max-count <max-count>
                     The maximum number of entries that can be stored in-memory by the
                       clientSessions cache. Available only when embedded Infinispan clusters
                       configured.
--cache-embedded-crl-max-count <max-count>
                     The maximum number of entries that can be stored in-memory by the crl cache.
--cache-embedded-keys-max-count <max-count>
                     The maximum number of entries that can be stored in-memory by the keys cache.
--cache-embedded-mtls-enabled <true|false>
                     Encrypts the network communication between Keycloak servers. Default: false.
--cache-embedded-mtls-key-store-file <file>
                     The Keystore file path. The Keystore must contain the certificate to use by
                       the TLS protocol. By default, it lookup 'cache-mtls-keystore.p12' under
                       conf/ directory.
...
```

So I would not add this point and the rest of the points for txt files, those files are for testing the command help displays.

andymunro commented on quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartDevHelpAll.approved.txt:
```diff
@@ -32,6 +32,8 @@ Cache:
                        clientSessions cache. Available only when embedded Infinispan clusters
                        configured.
+--cache-embedded-crl-max-count <max-count>
+                     The maximum number of entries that can be stored in-memory by the crl cache.
```
```suggestion
                     The maximum number of entries that can be stored in-memory by the crl cache
```

andymunro commented on quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartHelp.approved.txt:
```diff
@@ -33,6 +33,8 @@ Cache:
                        clientSessions cache. Available only when embedded Infinispan clusters
                        configured.
+--cache-embedded-crl-max-count <max-count>
+                     The maximum number of entries that can be stored in-memory by the crl cache.
```
```suggestion
                     The maximum number of entries that can be stored in-memory by the crl cache
```

andymunro commented on quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartHelpAll.approved.txt:
```diff
@@ -33,6 +33,8 @@ Cache:
                        clientSessions cache. Available only when embedded Infinispan clusters
                        configured.
+--cache-embedded-crl-max-count <max-count>
+                     The maximum number of entries that can be stored in-memory by the crl cache.
```
```suggestion
                     The maximum number of entries that can be stored in-memory by the crl cache
```

andymunro commented on quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartHelpAll.approved.txt:
```diff
@@ -507,4 +509,4 @@ By default, this command tries to update the server configuration by running a
 By doing that, the server should start faster based on any previous
-configuration you have set when manually running the 'build' command.
\ No newline at end of file
+configuration you have set when manually running the 'build' command.
```
```suggestion
configuration you have set when manually running the 'build' command
```

andymunro commented on quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartOptimizedHelp.approved.txt:
```diff
@@ -33,6 +33,8 @@ Cache:
                        clientSessions cache. Available only when embedded Infinispan clusters
                        configured.
+--cache-embedded-crl-max-count <max-count>
+                     The maximum number of entries that can be stored in-memory by the crl cache.
```
```suggestion
                     The maximum number of entries that can be stored in-memory by the crl cache
```

andymunro commented on quarkus/tests/integration/src/test/resources/org/keycloak/it/cli/dist/approvals/cli/help/HelpCommandDistTest.testStartOptimizedHelpAll.approved.txt:
```diff
@@ -33,6 +33,8 @@ Cache:
                        clientSessions cache. Available only when embedded Infinispan clusters
                        configured.
+--cache-embedded-crl-max-count <max-count>
+                     The maximum number of entries that can be stored in-memory by the crl cache.
```
```suggestion
                     The maximum number of entries that can be stored in-memory by the crl cache
```

andymunro reviewed (changes requested):
A few minor change, @rmartinc

rmartinc commented:
@mposolda @andymunro I have done the modifications suggested but not the ones in the txt files (command help displays). Those are generated programmatically and it means chaging all the caches, and I see all the options has a point in the end of each explanation. Moved the upgrading notes to relesae notes in 26.1.1.

mposolda reviewed (approved):
@rmartinc @pruivo @ahus1 @andymunro Thanks everyone for the updates and review!

mposolda commented:
@edewit Thanks!
